### PR TITLE
Allow invoking arbitrary user defined hooks

### LINKED ISSUE #79: Custom build-hooks not supported
We have a custom build-hook named `skms_cancel_cmr` defined in our .porter/config. When called in our Jenkins job it throws the following error.

```
21:40:41 INFO[09-06|21:40:41] Hook BEGIN                               porter_version=v2.1.1 cmd=build-hook HookName=skms_cancel_cmr
21:40:41 EROR[09-06|21:40:41] Invalid hook                             porter_version=v2.1.1 cmd=build-hook HookName=skms_cancel_cmr stack="[github.com/adobe-platform/porter/hook/hook.go:104 github.com/adobe-platform/porter/commands/build/hook.go:88 github.com/adobe-platform/porter/vendor/github.com/phylake/go-cli/cli.go:121 porter/main.go:36]"
```

Following the source breadcrumbs it appears custom hooks are not allowed due to this https://github.com/adobe-platform/porter/blob/master/hook/hook.go#L84-L106

### PR DESCRIPTION
**Pull request by jfieber**
## Changelog
- fix support for running abritrary user defined hooks
## Issues fixed or closed
#79
## Questions (open the PR then click the check boxes)

Did you update the documentation related to your changes?
- [x] Yes
- [ ] My changes were not already documented

Did you run `make` _before_ committing code and opening this PR?
- [x] Yes
- [ ] I didn't change any code

Did you run `porter create-stack` and `porter sync-stack` to verify provisioning
works?
- [x] Yes
- [ ] N/A

### CODE REVIEW

**Comment by jfieber** [comment]
@phylake, @gregoryfischer — ccweb stack is spinning, will check off the item when it finishes.  Currently posted `vJFIEBER` binary reflects this PR.

File: hook/hook.go [hunk 1]
```diff
@@ -81,27 +81,9 @@ func Execute(log log15.Logger,
...
-	case constants.HookPrePrune:
-		configHooks = config.Hooks.PrePrune
-	case constants.HookPostPrune:
-		configHooks = config.Hooks.PostPrune
-	case constants.HookEC2Bootstrap:
-		configHooks = config.Hooks.EC2Bootstrap
-	default:
-		log.Error("Invalid hook")
+	configHooks = config.Hooks[hookName]
+	if configHooks == nil {
```
Comment (phylake) [inline hook/hook.go]: This needs to be `configHooks, exists = config.Hooks[hookName]` and check for existence as a bool not configHooks as a pointer

File: docs/detailed_design/config-reference.md [hunk 2]
```diff
@@ -100,6 +100,12 @@ For each field the following notation is used
     - [dockerfile](#hook-dockerfile) (==1?)
     - [environment](#hook-environment) (==1?)
     - [concurrent](#concurrent) (==1?)
+  - user_defined (==1?)
```
Comment (phylake) [inline docs/detailed_design/config-reference.md]: Please add to the hook documentation describing this

File: conf/conf.go [hunk 3]
```diff
@@ -253,15 +241,9 @@ func (recv *Config) Print() {
...
 	fmt.Println(".Hooks")
-	printHooks(constants.HookPrePack, recv.Hooks.PrePack)
-	printHooks(constants.HookPostPack, recv.Hooks.PostPack)
-	printHooks(constants.HookPreProvision, recv.Hooks.PreProvision)
-	printHooks(constants.HookPostProvision, recv.Hooks.PostProvision)
-	printHooks(constants.HookPrePromote, recv.Hooks.PrePromote)
-	printHooks(constants.HookPostPromote, recv.Hooks.PostPromote)
-	printHooks(constants.HookPrePrune, recv.Hooks.PrePrune)
-	printHooks(constants.HookPostPrune, recv.Hooks.PostPrune)
-	printHooks(constants.HookEC2Bootstrap, recv.Hooks.EC2Bootstrap)
```
Comment (phylake) [inline conf/conf.go:264]: If these constants aren't in use elsewhere please remove them

Comment (jfieber) [reply conf/conf.go:264]: They are used elsewhere.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,7 @@
+**v2.2.1**
+
+- fix support for running abritrary user defined hooks
```
Comment (phylake) [inline CHANGELOG.md]: typo

**Comment by phylake** [comment]
👍 

[Approved with PullApprove](https://pullapprove.com/adobe-platform/porter/pull-request/97/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=adobe-platform/porter)

**Comment by jfieber** [comment]
:shipit: 

[Approved with PullApprove](https://pullapprove.com/adobe-platform/porter/pull-request/97/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=adobe-platform/porter)

**Pull request merged**